src/key_unwrapper.rs

```diff
@@ -0,0 +1,153 @@
+use std::collections::hash_map::Entry;
+use std::sync::{Arc, RwLock};
+
+/// Unwraps (decrypts) key encryption keys and data encryption keys using a KMS
+pub(crate) struct KeyUnwrapper {
+    kms_manager: Arc<KmsManager>,
+    kms_connection_config: RwLock<Arc<KmsConnectionConfig>>,
+    decryption_configuration: DecryptionConfiguration,
+    kek_cache: KekCache,
+}
+
+impl KeyUnwrapper {
+    pub fn new(
+        kms_manager: Arc<KmsManager>,
+        kms_connection_config: Arc<KmsConnectionConfig>,
+        decryption_configuration: DecryptionConfiguration,
+    ) -> Self {
+        let kek_cache = kms_manager.get_kek_cache(
+            &kms_connection_config,
+            decryption_configuration.cache_lifetime(),
+        );
+        let kms_connection_config = RwLock::new(kms_connection_config);
+        KeyUnwrapper {
+            kms_manager,
+            kms_connection_config,
+            decryption_configuration,
+            kek_cache,
+        }
+    }
+
+    fn unwrap_single_wrapped_key(&self, wrapped_dek: &str, master_key_id: &str) -> Result<Vec<u8>> {
+        let kms_connection_config = self.kms_connection_config.read().unwrap();
+        let client = self.kms_manager.get_client(
+            &kms_connection_config,
+            self.decryption_configuration.cache_lifetime(),
+        )?;
+        client.unwrap_key(wrapped_dek, master_key_id)
+    }
+
+    fn unwrap_double_wrapped_key(
+        &self,
+        wrapped_dek: &str,
+        master_key_id: &str,
+        kek_id: &str,
+        wrapped_kek: &str,
+    ) -> Result<Vec<u8>> {
+        let mut kek_cache = self.kek_cache.lock().unwrap();
+        let kek = match kek_cache.entry(kek_id.to_owned()) {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => {
+                let kms_connection_config = self.kms_connection_config.read().unwrap();
+                let client = self.kms_manager.get_client(
+                    &kms_connection_config,
+                    self.decryption_configuration.cache_lifetime(),
+                )?;
+                let kek = client.unwrap_key(wrapped_kek, master_key_id)?;
+                entry.insert(kek)
+            }
+        };
+        let decoded_kek_id = BASE64_STANDARD.decode(kek_id).map_err(|e| {
+            ParquetError::General(format!(
+                "Could not base64 decode key encryption key id: {}",
+                e
+            ))
+        })?;
+        key_encryption::decrypt_encryption_key(wrapped_dek, &decoded_kek_id, kek)
+    }
+
+    fn update_kms_config_from_footer_metadata(
+        &self,
+        kms_instance_id: &str,
+        kms_instance_url: &str,
+    ) -> Result<()> {
+        let mut kms_connection_config = self.kms_connection_config.write().unwrap();
+
+        if !kms_connection_config.kms_instance_id().is_empty()
+            && !kms_connection_config.kms_instance_url().is_empty()
+        {
+            return Ok(());
+        }
```

Comment:
If `!kms_connection_config.kms_instance_id().is_empty() && !kms_connection_config.kms_instance_url().is_empty()` we don't set the new `kms_instance_id` and `kms_instance_url`.
Is that desired behavior? (Might as well be, but then the name of method is misleading).
A test would be useful.

Reply:
Yes that's the desired behaviour. If those are set that means they've been specified by the user at read time so those settings should take precedence. There is a test for this here: https://github.com/G-Research/parquet-key-management-rs/blob/a6475ca4726a625848a918cc6c86c5b5a6f99d72/src/crypto_factory.rs#L739

I guess this could have been a bit clearer in the method name, or by adding a comment in the method. I'll follow up and tidy that up.

Reply:
Got it.